### PR TITLE
Added dev-container definition for dart+flutter+android

### LINKED ISSUE #162: Flutter dev-container
Please add a flutter container. See https://github.com/Dart-Code/Dart-Code/issues/2154
/cc @DanTup

### PR DESCRIPTION
**Pull request by MFAshby**
Adds a dev container for developing mobile apps with dart+flutter+android SDK

Resolves #162 

I copied the format of other containers for README.md, but I'm unsure what I need to add to the license section.

### CODE REVIEW

**Comment by bhack** [comment]
Could we use this devcontainer/Readme for in container android emulator?

**Comment by MFAshby** [comment]
Unknown, I have only tried it with physical devices. Additions welcome.

**Comment by bhack** [comment]
It could work with your image just with extra perm. See https://github.com/midnightblue69/docker-flutter/blob/master/README.md#run-it

**Comment by bhack** [comment]
Mhh the Google official launch script seems to me very similar to the other one I've already mentioned (X11, kvm device etc.). 
See https://github.com/google/android-emulator-container-scripts/blob/master/run-with-gpu.sh

**Comment by 741g** [comment]
Hi, thanks for the interest! What's your use case and the environment would you like to run it in? Our container scripts currently assume Linux due to the easier availability of virtualization (and, to our knowledge, Docker on macOS/Windows not supporting KVM on top of HVF/Hyper-V)

**Comment by 741g** [comment]
Also, what kind of feedback do you have for the container scripts themselves as far as a "better solution"? A standalone Dockerfile that doesn't require generation maybe? Please let us know :)

**Comment by bhack** [comment]
Was mainly about permissions but seems that we are on the same boat. I don't know if we need params for different emulator flavour.

File: containers/dart-flutter-android/.devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+	"name": "flutter",
```
Comment (Chuxel) [inline containers/dart-flutter-android/.devcontainer/devcontainer.json]: This is the name that will show in the lower left of the tool and in the dropdown in VS Code, so this isn't descriptive enough.  Perhaps "Dart & Flutter" for the name?

File: containers/dart-flutter-android/.devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+	"dockerFile": "Dockerfile",
+	"extensions": [
+		"dart-code.dart-code",
+		"dart-code.flutter"
+	],
+	"runArgs": ["--privileged", "-v", "/dev/bus/usb:/dev/bus/usb"]
```
[the remark this answers is not in the document]
Comment (MFAshby) [reply containers/dart-flutter-android/.devcontainer/devcontainer.json]: Either device access or an emulator is a necessity for this development environment. In my view interacting with `adb` installed on the host OS reduces the value of dockerizing the development environment: the docker container should encapsulate all the tools so I don't have to install lots of stuff on the host.

I'll have a look at options for interacting with USB devices on Windows & Mac OS hosts, but at first glance this seems to be non-trivial, so perhaps requiring `adb` on the host is the only way to make this cross-platform for now.

https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/hardware-devices-in-containers
https://devops.stackexchange.com/questions/4572/how-to-pass-a-dev-disk-device-on-macos-into-linux-docker

**Comment by bhack** [comment]
I agree with @MFAshby about usability. As @741g confirmed also for the emulator only  Google repo kvm has no solution to be exposed in Docker for Win/Osx https://github.com/microsoft/vscode-dev-containers/pull/169#issuecomment-566139493.
@Chuxel probably you want something like https://medium.com/@ccarcaci/android-emulation-on-docker-90d70ea95425 to be cross platform but I think it has not the usability that devs expect from a pre-packed container.

**Comment by bhack** [comment]
@Chuxel Can vscode detect host OS so that  It can list emulator/device access specilized dev-container only on Linux?

**Comment by bhack** [comment]
Just to reference the upstream issue for device acess:
https://github.com/docker/for-win/issues/3926
https://github.com/docker/for-mac/issues/900

For the emulator (X11/Kvm) It Is quite the same problem.

So IMHO the best option Is to have X11/KVM/USB Dev container for Linux (vscode can check the host OS to know what devcontainer offer in the list) and another sparated adb on host devcontainer (as the one I linked on the Medium

**Comment by MFAshby** [comment]
Addressed several comments @Chuxel , unfortunately not able to test until I've got a decent connection again but I hope to do this soon.

I still don't know what to do about cross-platform support. It seems unlikely that we'll get common USB device access on Windows & Mac. I also don't have a Windows or Mac OS machine available to test any changes.

**Comment by bhack** [comment]
Nobody here has control of OS "native" hypervisiors features. 
So in many cases IMHO It doesn't make sense to have unfied dev-containers.

It Is better to have host OS aware dev-container images lists. I don't know if we have an infra for this as vscode remote extension is currently closed source.

**Comment by Chuxel** [comment]
> Addressed several comments @Chuxel

@MFAshby Thanks!!

> It Is better to have host OS aware dev-container images lists. 

@bhack Unfortunately, this concept does not exist today since one of the value propositions of dev containers is also that teams that uses mixed operating systems can all benefit from a consistent toolchain. (This is therefore the focus of the containers in this repo.) That said, it could be added in concept for general use since I can see how that could be useful (even outside of this repo). I just raised https://github.com/microsoft/vscode-remote-release/issues/2067 to track this.

>I still don't know what to do about cross-platform support

@MFAshby @bhack Hmmmm.  If there really is no way to get this working outside of Linux at the moment and is of limited value without it, we'd likely need to implement something akin to what I mention in the issue above before we merged this in. I can see how it would be quite frustrating for someone to try the definition, hitting a failure, and after attempting to fix it coming to realize it will never work on their machine.

Now, something like the Flutter extension could opt to use the concept of a "helper extension" tactically to bridge operations that need to happen locally (and when combined with `asExternalUri`, can allow for file transfers from a container/SSH host to the local machine and forwarding of arbitrary ports programatically). However, that would still require that ADB and the android device drivers are on the host machine, so not sure if that really fixes the issue or not based on what you mention.

**Comment by bhack** [comment]
The main problem in cases like this one is that Docker itself it Is not fully features consistent across OS hypervisior backend technologies.
So as we are not in control of OSX and Microsoft hypervisiors and Vscode remote plugin (cause they are closed source) we really need to wait that MS will allocate Its own resource on https://github.com/microsoft/vscode-remote-release/issues/2067 in its monthly iteration.
For OSX, Win and Linux this Is quite ready https://medium.com/@ccarcaci/android-emulation-on-docker-90d70ea95425 so we could have a new PR based on that repo.

But I suggest to not lose this WIP while we are waiting for MS resource allocation on the upstream ticket.

@DanTup what do you think?

**Comment by pokowaka** [comment]
> https://medium.com/@ccarcaci/android-emulation-on-docker-90d70ea95425

Not so sure how you get it to work on Mac/Windows with Docker.  From my understanding docker on Mac is running in a hypervisor already and does not have nested viritualization and hence does not expose /dev/kvm.  The emulator will not run accelerated which will have a serious performance penalty (if it works at all).

I suspect that Docker on windows does not expose /dev/kvm either.

File: containers/dart-flutter-android/.devcontainer/devcontainer.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (pokowaka) [reply containers/dart-flutter-android/.devcontainer/devcontainer.json]: The --device flag does not work?  Note that adb in the container will not be able to interact with an emulator that runs outside of the container.

**Comment by bhack** [comment]
@pokowaka Yes seems it is using  https://github.com/budtmo/docker-android images so I suppose It Is not accelerated.

About nested virtualizzation on OSX the are the uptream tickets is:
https://github.com/moby/hyperkit/issues/127
I suppose that there is a similar ticket for Hyper-V in Docker for Win or Moby repo.

So the only crossplatform solution I see Is that the emulator Is running on host and you connect to it from the container.

But I suppose that in real projects where teams policy is full Dockerized probably they want to work with dev-container running on Linux host.

File: containers/dart-flutter-android/.devcontainer/devcontainer.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (MFAshby) [reply containers/dart-flutter-android/.devcontainer/devcontainer.json]: @pokowaka I believe that --device will not work, but I cannot try as I don't own a windows or mac machine to test it.

[the remark this answers is not in the document]
Comment (DanTup) [reply containers/dart-flutter-android/.devcontainer/devcontainer.json]: What exactly should the args look like to use `--device`? I can test it on Mac or Windows with a physical Android device.

File: containers/dart-flutter-android/.devcontainer/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ENV PATH="/android-sdk/tools/bin:/android-sdk/build-tools:/android-sdk/platform-tools:${PATH}"
+
+# SDK manager
+RUN yes | sdkmanager --licenses
+RUN sdkmanager "platforms;android-28" "platform-tools" "build-tools;28.0.3"
+
+# Flutter
+RUN wget https://storage.googleapis.com/flutter_infra/releases/stable/linux/flutter_linux_v1.5.4-hotfix.2-stable.tar.xz \
+ && tar xf flutter_linux_v1.5.4-hotfix.2-stable.tar.xz \
+ && rm flutter_linux_v1.5.4-hotfix.2-stable.tar.xz
```
Comment (DanTup) [inline containers/dart-flutter-android/.devcontainer/Dockerfile]: This seems like a fairly old version of Flutter? Current version is https://storage.googleapis.com/flutter_infra/releases/stable/linux/flutter_linux_v1.12.13+hotfix.5-stable.tar.xz

File: containers/dart-flutter-android/.devcontainer/devcontainer.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (MFAshby) [reply containers/dart-flutter-android/.devcontainer/devcontainer.json]: @DanTup not sure precisely, relevant docs are in here https://docs.docker.com/engine/reference/commandline/run/#add-host-device-to-container---device

**Comment by MFAshby** [comment]
I don't see a good solution to making this available cross-platform, and if that's a requirement for merging this PR then I don't think this will get merged. I'll close it for now;; those who wish to use it can still use it from my repo (or fork it)

**Comment by bhack** [comment]
I hope that we could have a good priority with  https://github.com/microsoft/vscode-remote-release/issues/2067.

**Pull request closed without merging**